Fix exact bit level checks leading to flakey tests

i.e.

```console
    --- FAIL: TestRustCoseCli/test_cose_sign_verify_P521_no_other_certs (0.10s)
        sign_verify_cose_rust_cli_test.go:56: 
            	Error Trace:	sign_verify_cose_rust_cli_test.go:56
            	            				sign_verify_cose_rust_cli_test.go:98
            	Error:      	Expected nil, but got: Byte lengths of integers r and s (9 and 8) do not match the key length 9
            	Test:       	TestRustCoseCli/test_cose_sign_verify_P521_no_other_certs
            	Messages:   	test_cose_sign_verify_P521_no_other_certs: signing failed with err Byte lengths of integers r and s (9 and 8) do not match the key length 9
        sign_verify_cose_rust_cli_test.go:91: 
            	Error Trace:	sign_verify_cose_rust_cli_test.go:91
            	            				sign_verify_cose_rust_cli_test.go:98
            	Error:      	Expected nil, but got: &exec.ExitError{ProcessState:(*os.ProcessState)(0xc00042e5c0), Stderr:[]uint8(nil)}
            	Test:       	TestRustCoseCli/test_cose_sign_verify_P521_no_other_certs
            	Messages:   	test_cose_sign_verify_P521_no_other_certs: error verifying signature with cose-rust exit status 2
```

from https://circleci.com/gh/mozilla-services/go-cose/23

--- a/core.go
+++ b/core.go
@@ -1,16 +1,16 @@
 package cose
 
 import (
-	"encoding/base64"
 	"crypto"
 	"crypto/ecdsa"
-	"crypto/rsa"
-	"crypto/rand"
 	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
 	"crypto/subtle"
+	"encoding/base64"
+	"github.com/pkg/errors"
 	"io"
 	"math/big"
-	"github.com/pkg/errors"
 )
 
 // ContextSignature identifies the context of the signature as a
@@ -89,12 +89,12 @@ func NewSigner(alg *Algorithm, options interface{}) (signer *Signer, err error) 
 			return nil, err
 		}
 	} else {
-                return nil, ErrUnknownPrivateKeyType
-        }
+		return nil, ErrUnknownPrivateKeyType
+	}
 
 	return &Signer{
 		PrivateKey: privateKey,
-		alg: alg,
+		alg:        alg,
 	}, nil
 }
 
@@ -109,7 +109,7 @@ func NewSignerFromKey(alg *Algorithm, privateKey crypto.PrivateKey) (signer *Sig
 	}
 	return &Signer{
 		PrivateKey: privateKey,
-		alg: alg,
+		alg:        alg,
 	}, nil
 }
 
@@ -304,9 +304,9 @@ func ecdsaCurveKeyBytesSize(curve elliptic.Curve) (keyBytesSize int) {
 // https://tools.ietf.org/html/rfc8017#section-4.1
 func I2OSP(b *big.Int, n int) []byte {
 	var (
-		octetString = b.Bytes()
+		octetString     = b.Bytes()
 		octetStringSize = len(octetString)
-		result = make([]byte, n)
+		result          = make([]byte, n)
 	)
 	if !(b.Sign() == 0 || b.Sign() == 1) {
 		panic("I2OSP error: integer must be zero or positive")
@@ -315,8 +315,8 @@ func I2OSP(b *big.Int, n int) []byte {
 		panic("I2OSP error: integer too large")
 	}
 
-	subtle.ConstantTimeCopy(1, result[:n - octetStringSize], result[:n - octetStringSize])
-	subtle.ConstantTimeCopy(1, result[n - octetStringSize:], octetString)
+	subtle.ConstantTimeCopy(1, result[:n-octetStringSize], result[:n-octetStringSize])
+	subtle.ConstantTimeCopy(1, result[n-octetStringSize:], octetString)
 	return result
 }
 
@@ -331,7 +331,6 @@ func FromBase64Int(data string) *big.Int {
 	}
 	return new(big.Int).SetBytes(val)
 }
-
 
 // Sign returns the SignatureBytes for each Signer in the same order
 // on the digest or the error from the first failing Signer

--- a/core.go
+++ b/core.go
@@ -158,9 +158,10 @@ func (s *Signer) Sign(rand io.Reader, digest []byte) (signature []byte, err erro
 		// These integers (r and s) will be the same length as
 		// the length of the key used for the signature
 		// process.
+		const tolerance = uint(1)
 		rByteLen, sByteLen, dByteLen := len(s.Bits()), len(r.Bits()), len(key.D.Bits())
-		if !(sByteLen == rByteLen && sByteLen == dByteLen) {
-			return nil, errors.Errorf("Byte lengths of integers r and s (%d and %d) do not match the key length %d\n", sByteLen, rByteLen, dByteLen)
+		if !(approxEqual(sByteLen, rByteLen, tolerance) && approxEqual(sByteLen, dByteLen, tolerance) && approxEqual(dByteLen, rByteLen, tolerance)) {
+			return nil, errors.Errorf("Byte lengths of integers r and s (%d and %d) do not match the key length %d±%d\n", sByteLen, rByteLen, dByteLen, tolerance)
 		}
 
 		// The signature is encoded by converting the integers
@@ -360,4 +361,20 @@ func Verify(digest []byte, signatures [][]byte, verifiers []ByteVerifier) (err e
 		}
 	}
 	return nil
+}
+
+// approxEquals returns a bool of whether x and y are equal to within
+// a given tolerance
+func approxEqual(x, y int, tolerance uint) bool {
+	var (
+		larger, smaller int
+	)
+	if x > y {
+		larger = x
+		smaller = y
+	} else {
+		larger = y
+		smaller = x
+	}
+	return uint(larger-smaller) <= tolerance
 }

--- a/core_test.go
+++ b/core_test.go
@@ -284,3 +284,16 @@ func TestI2OSPTiming(t *testing.T) {
 	fmt.Printf("I2OSPTiming timing diff is %s", time.Duration(diff))
 	assert.True(diff < toleranceNS)
 }
+
+func TestApproxEqual(t *testing.T) {
+	assert := assert.New(t)
+
+	assert.True(approxEqual(1, 1, 0))
+	assert.True(approxEqual(8, 9, 1))
+	assert.True(approxEqual(9, 8, 1))
+	assert.True(approxEqual(10, 5, 5))
+	assert.True(approxEqual(-1, 0, 1))
+
+	assert.False(approxEqual(10, 5, 1))
+	assert.False(approxEqual(6, 5, 0))
+}

